File: src/main/java/hu/gerviba/webschop/service/OpeningService.java
package hu.gerviba.webschop.service;

import java.text.SimpleDateFormat;
import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import hu.gerviba.webschop.dao.OpeningRepository;
import hu.gerviba.webschop.dao.TimeWindowRepository;
import hu.gerviba.webschop.model.OpeningEntity;
import hu.gerviba.webschop.model.TimeWindowEntity;

import javax.transaction.Transactional;

@Service
@Transactional
public class OpeningService {

    public static final SimpleDateFormat DATE_FORMATTER_HH_MM = new SimpleDateFormat("HH:mm"); 
    
    @Autowired
    OpeningRepository repo;

    @Autowired
    TimeWindowRepository twRepo;

    private static final long WEEK = 1000 * 60 * 60 * 24 * 7;
    
    public List<OpeningEntity> findAll() {
        return repo.findAllByOrderByDateStart();
    }

    public List<OpeningEntity> findUpcomingOpenings() {
        return repo.findAllByOrderEndGreaterThanOrderByDateStart(System.currentTimeMillis());
    }

    public List<OpeningEntity> findNextWeek() {
        return repo.findAllByDateEndGreaterThanAndDateEndLessThanOrderByDateStart(
                System.currentTimeMillis(), 
                System.currentTimeMillis() + WEEK);
    }
    
    public void save(OpeningEntity opening) {
        repo.save(opening);
    }
    
    public OpeningEntity findNextOf(Long id) {
        return repo.findFirstByCircle_IdAndDateEndGreaterThanOrderByDateStart(id, System.currentTimeMillis()).orElse(null);
    }
    
    public Long findNextStartDateOf(Long id) {
        Optional<OpeningEntity> opening = repo.findFirstByCircle_IdAndDateEndGreaterThanOrderByDateStart(id, System.currentTimeMillis());
        return opening.map(OpeningEntity::getDateStart).orElse(null);
    }

    public OpeningEntity getOne(Long openingId) {
        return repo.getOne(openingId);
    }

    public void delete(OpeningEntity oe) {
        repo.delete(oe);
    }

    public void saveTimeWindow(TimeWindowEntity tw) {
        twRepo.save(tw);
    }
    
}

File: src/main/kotlin/hu/kirdev/schpincer/service/OpeningService.kt
package hu.kirdev.schpincer.service

import hu.kirdev.schpincer.dao.OpeningRepository
import hu.kirdev.schpincer.dao.TimeWindowRepository
import hu.kirdev.schpincer.model.OpeningEntity
import hu.kirdev.schpincer.model.TimeWindowEntity
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.text.SimpleDateFormat
import java.time.Instant
import java.util.*

@Service
@Transactional
open class OpeningService {

    companion object {
        val DATE_FORMATTER_HH_MM by lazy { SimpleDateFormat("HH:mm") }
        private const val WEEK = 1000L * 60L * 60L * 24L * 7L
    }

    @Autowired
    private lateinit var repo: OpeningRepository

    @Autowired
    private lateinit var twRepo: TimeWindowRepository

    @Transactional(readOnly = true)
    open fun findAll(): List<OpeningEntity> {
        return repo.findAllByOrderByDateStart()
    }

    @Transactional(readOnly = true)
    open fun findUpcomingOpenings(): List<OpeningEntity> {
        return repo.findAllByOrderEndGreaterThanOrderByDateStart(Instant.now().toEpochMilli())
    }

    @Transactional(readOnly = true)
    open fun findNextWeek(): List<OpeningEntity> {
        return repo.findAllByDateEndGreaterThanAndDateEndLessThanOrderByDateStart(
                Instant.now().toEpochMilli(),
                Instant.now().toEpochMilli() + WEEK)
    }

    @Transactional(readOnly = true)
    open fun findEndedBefore(before: Long, limit: Long): List<OpeningEntity> {
        return repo.findAllEndedOpeningsBefore(Instant.now().toEpochMilli(), before, limit)
    }

    @Transactional(readOnly = false)
    open fun save(opening: OpeningEntity) {
        repo.save(opening)
    }

    @Transactional(readOnly = true)
    open fun findNextOf(id: Long): OpeningEntity? {
        return repo.findFirstByCircle_IdAndDateEndGreaterThanOrderByDateStart(id, Instant.now().toEpochMilli()).orElse(null)
    }

    @Transactional(readOnly = true)
    open fun findNextStartDateOf(id: Long): Long? {
        val opening: Optional<OpeningEntity> = repo.findFirstByCircle_IdAndDateEndGreaterThanOrderByDateStart(id, Instant.now().toEpochMilli())
        return opening.map(OpeningEntity::dateStart).orElse(null)
    }

    @Transactional(readOnly = true)
    open fun getOne(openingId: Long): OpeningEntity {
        return repo.getReferenceById(openingId)
    }

    @Transactional(readOnly = false)
    open fun delete(oe: OpeningEntity) {
        repo.delete(oe)
    }

    @Transactional(readOnly = false)
    open fun saveTimeWindow(tw: TimeWindowEntity) {
        twRepo.save(tw)
    }

    @Transactional(readOnly = true)
    open fun isCircleMatches(openingId: Long, circleId: Long): Boolean {
        return repo.findById(openingId).map { (it.circle?.id ?: 0) == circleId }.orElse(false)
    }

}
